synapse/rest/client/v2_alpha/room_keys.py

```diff
@@ -0,0 +1,279 @@
+                    "sessions": {
+                        "c0ff33": {
+                            "first_message_index": 1,
+                            "forwarded_count": 1,
+                            "is_verified": false,
+                            "session_data": "SSBBTSBBIEZJU0gK"
+                        }
+                    }
+                }
+            }
+        }
+        """
+        requester = yield self.auth.get_user_by_req(request, allow_guest=False)
+        user_id = requester.user.to_string()
+        version = request.args.get("version")[0]
+
+        room_keys = yield self.e2e_room_keys_handler.get_room_keys(
+            user_id, version, room_id, session_id
+        )
+        defer.returnValue((200, room_keys))
+
+    @defer.inlineCallbacks
+    def on_DELETE(self, request, room_id, session_id):
+        """
+        Deletes one or more encrypted E2E room keys for a user for backup purposes.
+
+        room_id: the ID of the room whose keys to delete (optional)
+        session_id: the ID for the E2E session to delete (optional)
+        version: the version of the user's backup which this data is for.
+        the version must already have been created via the /change_secret API.
+        """
+
+        requester = yield self.auth.get_user_by_req(request, allow_guest=False)
+        user_id = requester.user.to_string()
+        version = request.args.get("version")[0]
+
+        yield self.e2e_room_keys_handler.delete_room_keys(
+            user_id, version, room_id, session_id
+        )
+        defer.returnValue((200, {}))
+
+
+class RoomKeysVersionServlet(RestServlet):
+    PATTERNS = client_v2_patterns(
+        "/room_keys/version(/(?P<version>[^/]+))?$"
+    )
+
+    def __init__(self, hs):
+        """
+        Args:
+            hs (synapse.server.HomeServer): server
+        """
+        super(RoomKeysVersionServlet, self).__init__()
+        self.auth = hs.get_auth()
+        self.e2e_room_keys_handler = hs.get_e2e_room_keys_handler()
+
+    @defer.inlineCallbacks
+    def on_POST(self, request, version):
+        if version:
+            raise SynapseError(405, "Cannot POST to a specific version")
+
+        requester = yield self.auth.get_user_by_req(request, allow_guest=False)
+        user_id = requester.user.to_string()
+        info = parse_json_object_from_request(request)
+
+        new_version = yield self.e2e_room_keys_handler.create_version(
+            user_id, info
+        )
+        defer.returnValue((200, {"version": new_version}))
+
+    # we deliberately don't have a PUT /version, as these things really should
+    # be immutable to avoid people footgunning
+
+    @defer.inlineCallbacks
+    def on_GET(self, request, version):
+        requester = yield self.auth.get_user_by_req(request, allow_guest=False)
+        user_id = requester.user.to_string()
+
+        info = yield self.e2e_room_keys_handler.get_version_info(
+            user_id, version
+        )
```

Comment:
Should we be checking that version was actually supplied? Would it make more sense to spit this class in two, one doing `POST /room_keys/version` and the other `GET/DELETE /room_keys/version/<version>`?

In fact, having POST <verb> might makes things a bit clearer, e.g. `POST /room_keys/create_version`

Reply:
I didn't bother checking that version was supplied; i think it will explode fairly sensibly if it isn't?

Not sure what the point of splitting into POST and GET/DELETE would be?  And shouldn't POST have the same path anyway? (although ugh rest https://medium.com/@pakaldebonchamp/rest-is-the-new-soap-97ff6c09896d)

Reply:
> I didn't bother checking that version was supplied; i think it will explode fairly sensibly if it isn't?

My hunch is that it'd 500 in a fairly non obvious way tbh, or possibly succeed as the database ends up executing `SELECT * FROM versions WHERE version = null`. Not hugely friendly.

> Not sure what the point of splitting into POST and GET/DELETE would be?

So you have a GET and DELETE that only work if a version is supplied, and a POST that only works if you *don't* supply a version. To save having to do checks everywhere, and to make it clearer when a version is required or disallowed, having this as separate classes seems much clearer.

> And shouldn't POST have the same path anyway? 

They don't really share a path as it is. If you think that `POST /room_keys/version` is clearer than `POST /room_keys/create_version` then fair enough, twas but a thought.

And yes, turns out naming APIs is hard.

Reply:
I've changed the semantics so that if version is missing it gives you the most recent version (and 404 if there are none).

I've experimented with building /version into the path, but it feels pretty clunky (given you'd need to know the version in order to know the path, meaning that things like 'give me the current version' are harder to express).

I've also thought about whether `POST /room_keys/version` returning an ID which can then be used with `GET/DELETE /room_keys/version/$id` is bad form, but from my pov this is a pretty standard RESTy pattern.  Especially as we now also have `GET /room_keys/version` to return the current version.  Also, I think this is increasingly dubiously subjective, and in practice as long as it works I'm not sure the precise API shape is that critical unless it's spectacularly inconsistent with the wider CS API.

Reply:
Honestly, the thought of changing the API shape was simply a tangential suggestion. The actual suggestion was to change the code layout to better avoid having to manually handle null values of `version`.

Having `GET /version` that returns the most recent version works, but what does it return if there isn't a version for that user? What does `DELETE /version` do? I note that we still don't seem to be checking if the version is null in that case.

Reply:
for `GET /version` when there is no most recent version, you should get a 404, as the docstring says.

I've fixed `DELETE /version` to delete the current backup version info, for symmetry with the rest of the API, and added a test for it.